#!/usr/bin/env python3
"""
–ü–æ–ª–Ω—ã–π E2E —Ç–µ—Å—Ç –æ–±—Ä–∞–±–æ—Ç–∫–∏ –∑–∞–ø—Ä–æ—Å–∞ –æ—Ç –∫–ª–∏–µ–Ω—Ç–∞ –¥–æ —Å–µ—Ä–≤–µ—Ä–∞

–¢–µ—Å—Ç–∏—Ä—É–µ—Ç:
1. –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ gRPC —Å–µ—Ä–≤–µ—Ä—É
2. –û—Ç–ø—Ä–∞–≤–∫–∞ StreamRequest
3. –ü–æ–ª—É—á–µ–Ω–∏–µ –ø–æ—Ç–æ–∫–æ–≤—ã—Ö –æ—Ç–≤–µ—Ç–æ–≤ (text_chunk, audio_chunk)
4. –ò–∑–º–µ—Ä–µ–Ω–∏–µ –≤—Ä–µ–º–µ–Ω–∏ –Ω–∞ –∫–∞–∂–¥–æ–º —ç—Ç–∞–ø–µ
5. –í–∞–ª–∏–¥–∞—Ü–∏—è –≤—Å–µ—Ö —á–∞–Ω–∫–æ–≤
"""

import asyncio
import sys
import time
from pathlib import Path
from typing import List, Dict, Any
import grpc
import grpc.aio as aio

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ –º–æ–¥—É–ª—è–º
project_root = Path(__file__).parent.parent
sys.path.insert(0, str(project_root))

# –ò–º–ø–æ—Ä—Ç–∏—Ä—É–µ–º protobuf —Ñ–∞–π–ª—ã
try:
    from modules.grpc_service import streaming_pb2, streaming_pb2_grpc
except ImportError:
    # –ê–ª—å—Ç–µ—Ä–Ω–∞—Ç–∏–≤–Ω—ã–π –ø—É—Ç—å –¥–ª—è –∑–∞–ø—É—Å–∫–∞ –∏–∑ –∫–æ—Ä–Ω—è –ø—Ä–æ–µ–∫—Ç–∞
    server_root = Path(__file__).parent.parent.parent
    sys.path.insert(0, str(server_root))
    from server.modules.grpc_service import streaming_pb2, streaming_pb2_grpc


class RequestFlowTester:
    """–¢–µ—Å—Ç–µ—Ä –ø–æ–ª–Ω–æ–≥–æ –ø—É—Ç–∏ –æ–±—Ä–∞–±–æ—Ç–∫–∏ –∑–∞–ø—Ä–æ—Å–∞"""
    
    def __init__(self, host: str = "localhost", port: int = 50051):
        self.host = host
        self.port = port
        self.address = f"{host}:{port}"
        
        # –ú–µ—Ç—Ä–∏–∫–∏
        self.metrics = {
            'connection_time': 0,
            'first_text_chunk_time': 0,
            'first_audio_chunk_time': 0,
            'total_time': 0,
            'text_chunks_count': 0,
            'audio_chunks_count': 0,
            'total_audio_bytes': 0,
            'errors': []
        }
        
    async def test_full_flow(self, prompt: str = "Hello, can you help me?", hardware_id: str = "test_hardware_123") -> Dict[str, Any]:
        """
        –ü–æ–ª–Ω—ã–π —Ç–µ—Å—Ç –æ–±—Ä–∞–±–æ—Ç–∫–∏ –∑–∞–ø—Ä–æ—Å–∞
        
        Args:
            prompt: –¢–µ–∫—Å—Ç –∑–∞–ø—Ä–æ—Å–∞
            hardware_id: –ò–¥–µ–Ω—Ç–∏—Ñ–∏–∫–∞—Ç–æ—Ä –æ–±–æ—Ä—É–¥–æ–≤–∞–Ω–∏—è
            
        Returns:
            –°–ª–æ–≤–∞—Ä—å —Å —Ä–µ–∑—É–ª—å—Ç–∞—Ç–∞–º–∏ —Ç–µ—Å—Ç–∞
        """
        print("=" * 80)
        print("üß™ –ü–û–õ–ù–´–ô E2E –¢–ï–°–¢ –û–ë–†–ê–ë–û–¢–ö–ò –ó–ê–ü–†–û–°–ê")
        print("=" * 80)
        print(f"\nüì§ –ó–∞–ø—Ä–æ—Å:")
        print(f"   Prompt: {prompt}")
        print(f"   Hardware ID: {hardware_id}")
        print(f"   Server: {self.address}")
        print()
        
        start_time = time.time()
        
        try:
            # –≠—Ç–∞–ø 1: –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ
            print("üîå –≠—Ç–∞–ø 1: –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ —Å–µ—Ä–≤–µ—Ä—É...")
            connection_start = time.time()
            
            channel = aio.insecure_channel(self.address)
            stub = streaming_pb2_grpc.StreamingServiceStub(channel)
            
            try:
                await asyncio.wait_for(channel.channel_ready(), timeout=5.0)
                connection_time = time.time() - connection_start
                self.metrics['connection_time'] = connection_time * 1000
                print(f"   ‚úÖ –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ —É—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω–æ –∑–∞ {connection_time*1000:.2f}ms")
            except asyncio.TimeoutError:
                print("   ‚ùå –¢–∞–π–º–∞—É—Ç –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è")
                return {'success': False, 'error': 'Connection timeout', 'metrics': self.metrics}
            
            # –≠—Ç–∞–ø 2: –°–æ–∑–¥–∞–Ω–∏–µ –∑–∞–ø—Ä–æ—Å–∞
            print("\nüìù –≠—Ç–∞–ø 2: –°–æ–∑–¥–∞–Ω–∏–µ StreamRequest...")
            request = streaming_pb2.StreamRequest(
                prompt=prompt,
                hardware_id=hardware_id,
                session_id=f"test_session_{int(time.time())}"
            )
            print(f"   ‚úÖ –ó–∞–ø—Ä–æ—Å —Å–æ–∑–¥–∞–Ω: session_id={request.session_id}")
            
            # –≠—Ç–∞–ø 3: –û—Ç–ø—Ä–∞–≤–∫–∞ –∑–∞–ø—Ä–æ—Å–∞ –∏ –ø–æ–ª—É—á–µ–Ω–∏–µ –æ—Ç–≤–µ—Ç–æ–≤
            print("\nüîÑ –≠—Ç–∞–ø 3: –û—Ç–ø—Ä–∞–≤–∫–∞ –∑–∞–ø—Ä–æ—Å–∞ –∏ –ø–æ–ª—É—á–µ–Ω–∏–µ –ø–æ—Ç–æ–∫–æ–≤—ã—Ö –æ—Ç–≤–µ—Ç–æ–≤...")
            print("   (–û–∂–∏–¥–∞–Ω–∏–µ –æ—Ç–≤–µ—Ç–æ–≤ –æ—Ç —Å–µ—Ä–≤–µ—Ä–∞...)\n")
            
            request_start = time.time()
            first_text_time = None
            first_audio_time = None
            
            text_chunks: List[str] = []
            audio_chunks: List[bytes] = []
            end_message = None
            error_message = None
            
            try:
                async for response in stub.StreamAudio(request, timeout=60.0):
                    response_time = time.time() - request_start
                    content_type = response.WhichOneof("content")
                    
                    if content_type == "text_chunk":
                        if first_text_time is None:
                            first_text_time = response_time
                            self.metrics['first_text_chunk_time'] = first_text_time * 1000
                            print(f"   üìù –ü–µ—Ä–≤—ã–π text_chunk –ø–æ–ª—É—á–µ–Ω —á–µ—Ä–µ–∑ {first_text_time*1000:.2f}ms")
                        
                        text = response.text_chunk
                        text_chunks.append(text)
                        self.metrics['text_chunks_count'] += 1
                        print(f"   üìù Text chunk #{len(text_chunks)}: '{text[:50]}{'...' if len(text) > 50 else ''}' ({len(text)} —Å–∏–º–≤–æ–ª–æ–≤)")
                    
                    elif content_type == "audio_chunk":
                        if first_audio_time is None:
                            first_audio_time = response_time
                            self.metrics['first_audio_chunk_time'] = first_audio_time * 1000
                            print(f"   üéµ –ü–µ—Ä–≤—ã–π audio_chunk –ø–æ–ª—É—á–µ–Ω —á–µ—Ä–µ–∑ {first_audio_time*1000:.2f}ms")
                        
                        audio = response.audio_chunk
                        audio_data = audio.audio_data
                        audio_chunks.append(audio_data)
                        self.metrics['audio_chunks_count'] += 1
                        self.metrics['total_audio_bytes'] += len(audio_data)
                        
                        print(f"   üéµ Audio chunk #{len(audio_chunks)}: {len(audio_data)} bytes "
                              f"(sample_rate={audio.sample_rate}, channels={audio.channels}, dtype={audio.dtype})")
                    
                    elif content_type == "end_message":
                        end_message = response.end_message
                        total_time = time.time() - request_start
                        self.metrics['total_time'] = total_time * 1000
                        print(f"\n   ‚úÖ End message –ø–æ–ª—É—á–µ–Ω —á–µ—Ä–µ–∑ {total_time*1000:.2f}ms: '{end_message}'")
                        break
                    
                    elif content_type == "error_message":
                        error_message = response.error_message
                        print(f"\n   ‚ùå Error message: '{error_message}'")
                        self.metrics['errors'].append(error_message)
                        break
                
                # –ï—Å–ª–∏ –Ω–µ –±—ã–ª–æ end_message –∏–ª–∏ error_message, –Ω–æ —Å—Ç—Ä–∏–º –∑–∞–∫–æ–Ω—á–∏–ª—Å—è
                if end_message is None and error_message is None:
                    total_time = time.time() - request_start
                    self.metrics['total_time'] = total_time * 1000
                    print(f"\n   ‚ö†Ô∏è –°—Ç—Ä–∏–º –∑–∞–≤–µ—Ä—à–∏–ª—Å—è –±–µ–∑ end_message –∏–ª–∏ error_message (–≤—Ä–µ–º—è: {total_time*1000:.2f}ms)")
                
            except grpc.RpcError as e:
                error_time = time.time() - request_start
                error_msg = f"gRPC –æ—à–∏–±–∫–∞: {e.code()} - {e.details()}"
                print(f"\n   ‚ùå {error_msg} (–≤—Ä–µ–º—è: {error_time*1000:.2f}ms)")
                self.metrics['errors'].append(error_msg)
                self.metrics['total_time'] = error_time * 1000
                return {'success': False, 'error': error_msg, 'metrics': self.metrics}
            
            except asyncio.TimeoutError:
                error_time = time.time() - request_start
                error_msg = "–¢–∞–π–º–∞—É—Ç –æ–∂–∏–¥–∞–Ω–∏—è –æ—Ç–≤–µ—Ç–æ–≤ (60 —Å–µ–∫—É–Ω–¥)"
                print(f"\n   ‚ùå {error_msg} (–≤—Ä–µ–º—è: {error_time*1000:.2f}ms)")
                self.metrics['errors'].append(error_msg)
                self.metrics['total_time'] = error_time * 1000
                return {'success': False, 'error': error_msg, 'metrics': self.metrics}
            
            finally:
                await channel.close()
            
            # –≠—Ç–∞–ø 4: –ê–Ω–∞–ª–∏–∑ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤
            print("\n" + "=" * 80)
            print("üìä –†–ï–ó–£–õ–¨–¢–ê–¢–´ –¢–ï–°–¢–ê")
            print("=" * 80)
            
            success = error_message is None and len(text_chunks) > 0
            
            print(f"\n‚úÖ –£—Å–ø–µ—à–Ω–æ—Å—Ç—å: {'‚úÖ –£–°–ü–ï–•' if success else '‚ùå –û–®–ò–ë–ö–ê'}")
            print(f"\nüìà –ú–µ—Ç—Ä–∏–∫–∏:")
            print(f"   ‚Ä¢ –í—Ä–µ–º—è –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è: {self.metrics['connection_time']:.2f}ms")
            print(f"   ‚Ä¢ –í—Ä–µ–º—è –¥–æ –ø–µ—Ä–≤–æ–≥–æ text_chunk: {self.metrics['first_text_chunk_time']:.2f}ms")
            print(f"   ‚Ä¢ –í—Ä–µ–º—è –¥–æ –ø–µ—Ä–≤–æ–≥–æ audio_chunk: {self.metrics['first_audio_chunk_time']:.2f}ms")
            print(f"   ‚Ä¢ –û–±—â–µ–µ –≤—Ä–µ–º—è –æ–±—Ä–∞–±–æ—Ç–∫–∏: {self.metrics['total_time']:.2f}ms ({self.metrics['total_time']/1000:.2f} —Å–µ–∫)")
            print(f"\nüì¶ –ü–æ–ª—É—á–µ–Ω–Ω—ã–µ –¥–∞–Ω–Ω—ã–µ:")
            print(f"   ‚Ä¢ Text chunks: {self.metrics['text_chunks_count']}")
            print(f"   ‚Ä¢ Audio chunks: {self.metrics['audio_chunks_count']}")
            print(f"   ‚Ä¢ –í—Å–µ–≥–æ –∞—É–¥–∏–æ –¥–∞–Ω–Ω—ã—Ö: {self.metrics['total_audio_bytes']} bytes ({self.metrics['total_audio_bytes']/1024:.2f} KB)")
            
            if text_chunks:
                full_text = " ".join(text_chunks)
                print(f"\nüìù –ü–æ–ª–Ω—ã–π —Ç–µ–∫—Å—Ç –æ—Ç–≤–µ—Ç–∞:")
                print(f"   '{full_text[:200]}{'...' if len(full_text) > 200 else ''}'")
                print(f"   (–í—Å–µ–≥–æ —Å–∏–º–≤–æ–ª–æ–≤: {len(full_text)})")
            
            if self.metrics['errors']:
                print(f"\n‚ùå –û—à–∏–±–∫–∏:")
                for error in self.metrics['errors']:
                    print(f"   ‚Ä¢ {error}")
            
            # –ê–Ω–∞–ª–∏–∑ –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç–µ–ª—å–Ω–æ—Å—Ç–∏
            print(f"\n‚ö° –ê–Ω–∞–ª–∏–∑ –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç–µ–ª—å–Ω–æ—Å—Ç–∏:")
            if self.metrics['first_text_chunk_time'] > 0:
                print(f"   ‚Ä¢ LLM –æ–±—Ä–∞–±–æ—Ç–∫–∞ (–¥–æ –ø–µ—Ä–≤–æ–≥–æ text): {self.metrics['first_text_chunk_time']:.2f}ms")
            if self.metrics['first_audio_chunk_time'] > 0:
                tts_time = self.metrics['first_audio_chunk_time'] - self.metrics['first_text_chunk_time']
                if tts_time > 0:
                    print(f"   ‚Ä¢ TTS –æ–±—Ä–∞–±–æ—Ç–∫–∞ (text ‚Üí audio): {tts_time:.2f}ms")
            
            total_elapsed = time.time() - start_time
            print(f"\n‚è±Ô∏è –û–±—â–µ–µ –≤—Ä–µ–º—è —Ç–µ—Å—Ç–∞: {total_elapsed*1000:.2f}ms ({total_elapsed:.2f} —Å–µ–∫)")
            
            return {
                'success': success,
                'text_chunks': text_chunks,
                'audio_chunks_count': len(audio_chunks),
                'total_audio_bytes': self.metrics['total_audio_bytes'],
                'end_message': end_message,
                'error_message': error_message,
                'metrics': self.metrics
            }
            
        except Exception as e:
            error_msg = f"–ù–µ–æ–∂–∏–¥–∞–Ω–Ω–∞—è –æ—à–∏–±–∫–∞: {str(e)}"
            print(f"\n‚ùå {error_msg}")
            import traceback
            traceback.print_exc()
            self.metrics['errors'].append(error_msg)
            return {'success': False, 'error': error_msg, 'metrics': self.metrics}


async def main():
    """–ì–ª–∞–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è"""
    import argparse
    
    parser = argparse.ArgumentParser(description='E2E —Ç–µ—Å—Ç –ø–æ–ª–Ω–æ–≥–æ –ø—É—Ç–∏ –æ–±—Ä–∞–±–æ—Ç–∫–∏ –∑–∞–ø—Ä–æ—Å–∞')
    parser.add_argument('--host', default='localhost', help='–ê–¥—Ä–µ—Å —Å–µ—Ä–≤–µ—Ä–∞ (–ø–æ —É–º–æ–ª—á–∞–Ω–∏—é: localhost)')
    parser.add_argument('--port', type=int, default=50051, help='–ü–æ—Ä—Ç —Å–µ—Ä–≤–µ—Ä–∞ (–ø–æ —É–º–æ–ª—á–∞–Ω–∏—é: 50051)')
    parser.add_argument('--prompt', default='Hello, can you help me?', help='–¢–µ–∫—Å—Ç –∑–∞–ø—Ä–æ—Å–∞')
    parser.add_argument('--hardware-id', default='test_hardware_123', help='Hardware ID')
    
    args = parser.parse_args()
    
    tester = RequestFlowTester(host=args.host, port=args.port)
    result = await tester.test_full_flow(prompt=args.prompt, hardware_id=args.hardware_id)
    
    # –í–æ–∑–≤—Ä–∞—â–∞–µ–º –∫–æ–¥ –≤—ã—Ö–æ–¥–∞
    sys.exit(0 if result.get('success') else 1)


if __name__ == "__main__":
    asyncio.run(main())
